File: algorithms/02_c/selection/quickselect.c
/* Quickselect: expected linear-time order statistic (CLRS 9.2). */
#include <assert.h>
#include <stdio.h>

static int quickselect(int *a, int n, int k) {
    int lo = 0, hi = n - 1;
    while (1) {
        int pivot = a[hi], i = lo, j;
        for (j = lo; j < hi; ++j) {
            if (a[j] < pivot) {
                int t = a[i]; a[i++] = a[j]; a[j] = t;
            }
        }
        int t = a[i]; a[i] = a[hi]; a[hi] = t;
        if (i == k) return a[i];
        if (k < i) hi = i - 1; else lo = i + 1;
    }
}

int main(void) {
    int data[6] = {3, 2, 1, 5, 6, 4};
    int expected[6] = {1, 2, 3, 4, 5, 6};
    for (int k = 0; k < 6; ++k) {
        int copy[6];
        for (int i = 0; i < 6; ++i) copy[i] = data[i];
        assert(quickselect(copy, 6, k) == expected[k]);
    }
    printf("quickselect ok\n");
    return 0;
}

File: algorithms/06_java/selection/Quickselect.java
import java.util.Arrays;

// Quickselect: expected linear-time order statistic (CLRS 9.2).
public class Quickselect {
    static int quickselect(int[] a, int k) {
        int lo = 0, hi = a.length - 1;
        while (true) {
            int pivot = a[hi], i = lo;
            for (int j = lo; j < hi; j++) {
                if (a[j] < pivot) {
                    int t = a[i]; a[i++] = a[j]; a[j] = t;
                }
            }
            int t = a[i]; a[i] = a[hi]; a[hi] = t;
            if (i == k) return a[i];
            if (k < i) hi = i - 1; else lo = i + 1;
        }
    }

    public static void main(String[] args) {
        int[] data = {3, 2, 1, 5, 6, 4};
        int[] sorted = data.clone();
        Arrays.sort(sorted);
        for (int k = 0; k < data.length; k++) {
            assert quickselect(data.clone(), k) == sorted[k];
        }
        System.out.println("quickselect ok");
    }
}
